File: nuxt-app/server/utils/rate-limit.ts
interface RateLimitEntry {
  count: number
  resetAt: number
}

const buckets = new Map<string, RateLimitEntry>()

function sweep(now: number) {
  if (buckets.size < 1000) {
    return
  }
  for (const [key, entry] of buckets) {
    if (entry.resetAt <= now) {
      buckets.delete(key)
    }
  }
}

export function checkRateLimit(key: string, limit: number, windowMs: number): boolean {
  const now = Date.now()
  sweep(now)

  const entry = buckets.get(key)
  if (!entry || entry.resetAt <= now) {
    buckets.set(key, { count: 1, resetAt: now + windowMs })
    return true
  }

  if (entry.count >= limit) {
    return false
  }

  entry.count++
  return true
}

File: rust-server/src/rate_limit.rs
use std::collections::HashMap;
use std::sync::Mutex;
use std::time::{Duration, Instant};

struct Entry {
    count: u32,
    reset_at: Instant,
}

#[derive(Default)]
pub struct RateLimiter {
    buckets: Mutex<HashMap<String, Entry>>,
}

impl RateLimiter {
    pub fn check(&self, key: &str, limit: u32, window: Duration) -> bool {
        let mut buckets = self.buckets.lock().unwrap();
        let now = Instant::now();

        if buckets.len() >= 1000 {
            buckets.retain(|_, e| e.reset_at > now);
        }

        match buckets.get_mut(key) {
            Some(entry) if entry.reset_at > now => {
                if entry.count >= limit {
                    false
                } else {
                    entry.count += 1;
                    true
                }
            }
            _ => {
                buckets.insert(
                    key.to_string(),
                    Entry {
                        count: 1,
                        reset_at: now + window,
                    },
                );
                true
            }
        }
    }
}
